spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindableProxyFactory.java

```diff
@@ -0,0 +1,316 @@
+							final MessageChannel outputChannel = createMessageChannel(messageChannelType);
+							if (isPollable(messageChannelType)) {
+								bridgePollableToSubscribableChannel(outputChannel,
+										sharedChannel);
+							}
+							else {
+								bridgeSubscribableToPollableChannel(
+										(SubscribableChannel) outputChannel,
+										sharedChannel);
+							}
+							outputs.put(name, new ChannelHolder(outputChannel, false));
+						}
+					}
+				}
+			}
+		});
+	}
+
+	private MessageChannel locateSharedChannel(String name) {
+		return sharedChannelRegistry != null ? sharedChannelRegistry
+				.getSharedChannel(getNamespacePrefixedChannelName(name)) : null;
+	}
+
+	private String getNamespacePrefixedChannelName(String name) {
+		return channelNamespace + "." + name;
+	}
+
+	private void bridgeSubscribableToPollableChannel(SubscribableChannel sharedChannel,
+			MessageChannel inputChannel) {
+		sharedChannel.subscribe(new MessageChannelBinderSupport.DirectHandler(
+				inputChannel));
+	}
+
+	private void bridgePollableToSubscribableChannel(MessageChannel pollableChannel,
+			MessageChannel subscribableChannel) {
+		ConsumerEndpointFactoryBean consumerEndpointFactoryBean = new ConsumerEndpointFactoryBean();
+		consumerEndpointFactoryBean.setInputChannel(pollableChannel);
+		PollerMetadata pollerMetadata = new PollerMetadata();
+		pollerMetadata.setTrigger(new PeriodicTrigger(1000));
+		consumerEndpointFactoryBean.setPollerMetadata(pollerMetadata);
+		consumerEndpointFactoryBean
+				.setHandler(new MessageChannelBinderSupport.DirectHandler(
+						subscribableChannel));
+		consumerEndpointFactoryBean.setBeanFactory(beanFactory);
+		try {
+			consumerEndpointFactoryBean.afterPropertiesSet();
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+		}
+		consumerEndpointFactoryBean.start();
+	}
+
+	private MessageChannel createMessageChannel(Class<?> messageChannelType) {
+		return isPollable(messageChannelType) ? new QueueChannel() : new DirectChannel();
+	}
+
+	private boolean isPollable(Class<?> channelType) {
+		return PollableChannel.class.equals(channelType);
+	}
+
+	@Override
+	public synchronized Object invoke(MethodInvocation invocation) throws Throwable {
+		Method method = invocation.getMethod();
+		if (MessageChannel.class.isAssignableFrom(method.getReturnType())) {
+			Input input = AnnotationUtils.findAnnotation(method, Input.class);
+			if (input != null) {
+				String name = BindingBeanDefinitionRegistryUtils.getChannelName(input,
+						method);
+				return this.inputs.get(name).getMessageChannel();
+			}
+			Output output = AnnotationUtils.findAnnotation(method, Output.class);
+			if (output != null) {
+				String name = BindingBeanDefinitionRegistryUtils.getChannelName(output,
+						method);
+				return this.outputs.get(name).getMessageChannel();
+			}
+		}
+		// ignore
+		return null;
+	}
+
+	@Override
+	public synchronized Object getObject() throws Exception {
```

Comment:
Would it make sense to separate the Proxy implementation from the `FactoryBean` implementation?